### PR TITLE
Interpolation

### PR DESCRIPTION
**Pull request by purple-waves**
Implement very basic interpolation strategy: Apply the aqi to neighbouring cells with a decreasing weighting.

Deployed version: https://interpolation-dot-step269-2020.ts.r.appspot.com

### CODE REVIEW

File: src/main/java/com/google/sps/data/Cache.java [hunk 1]
```diff
@@ -28,16 +28,33 @@ public GriddedData getGrid(Coordinates swCorner, Coordinates neCorner) {
...
-      GridCell cell = row.getOrDefault(index.col, new GridCell());
-      cell.addPoint(dataPoint);
-      row.put(index.col, cell);
-      dataGrid.put(index.row, row);
+      addDataPointWithWeighting(dataPoint, 5, index);
+      // HashMap<Integer,GridCell> row = dataGrid.getOrDefault(index.row, new HashMap<>());
+      // GridCell cell = row.getOrDefault(index.col, new GridCell());
+      // cell.addPoint(dataPoint);
+      // row.put(index.col, cell);
+      // dataGrid.put(index.row, row);
```
Comment (roroross) [inline src/main/java/com/google/sps/data/Cache.java]: commented out code?

File: src/main/java/com/google/sps/data/Cache.java [hunk 2]
```diff
@@ -28,16 +28,33 @@ public GriddedData getGrid(Coordinates swCorner, Coordinates neCorner) {
...
 
     return convertToGriddedData(swCorner, neCorner);
   }
 
+  private void addDataPointWithWeighting(AQDataPoint dataPoint, int numRings, GridIndex centreCellIndex) {
+    int centreRow = centreCellIndex.row;
+    int centreCol = centreCellIndex.col;
+    for (int ring = numRings -1; ring >= 0; ring--) {
+      for (int rowNum = centreRow - ring; rowNum <= centreRow + ring; rowNum ++) {
+        HashMap<Integer,GridCell> row = dataGrid.getOrDefault(rowNum, new HashMap<>());
```
Comment (roroross) [inline src/main/java/com/google/sps/data/Cache.java]: spacing in hashmap

File: src/main/java/com/google/sps/data/EmptyListException.java [hunk 3]
```diff
@@ -4,6 +4,8 @@
  * A Custom Exception class for an Empty List Exception. 
  */
 public class EmptyListException extends Exception {
+    private static final long serialVersionUID = 6243104809253949369L;
```
Comment (roroross) [inline src/main/java/com/google/sps/data/EmptyListException.java:7]: what is this number used for / where is it from?

File: src/main/webapp/script.js [hunk 4]
```diff
@@ -103,7 +103,7 @@ function loadHeatmap(data) {
             ],
             maxIntensity: 200,
             dissipating: true,
-            radius: 100, // arbitrary value used to make the points easily visible
+            radius: 10, // arbitrary value
```
Comment (roroross) [inline src/main/webapp/script.js]: capital and full stop

File: src/main/webapp/script.js [hunk 5]
```diff
@@ -112,6 +112,7 @@ function loadHeatmap(data) {
 
 function convertGriddedDataToWeightedPoints(griddedData) {
     const dataGrid = griddedData.data;
+    console.log(dataGrid);
```
Comment (roroross) [inline src/main/webapp/script.js]: ifs this for debugging? or part of code

Comment (roroross) [reply src/main/webapp/script.js]: actually, may be useful for debugging on my end too? 😅 but i meant will this be here for the production code?

File: src/main/java/com/google/sps/data/Cache.java (hunk 1 above)
Comment (purple-waves) [reply src/main/java/com/google/sps/data/Cache.java]: Gahhh where was my brain at yesterday

File: src/main/java/com/google/sps/data/EmptyListException.java (hunk 3 above)
Comment (purple-waves) [reply src/main/java/com/google/sps/data/EmptyListException.java:7]: It was to get rid of a warning that wanted it to have a unique id, the number was generated randomly by a suggested fix

File: src/main/webapp/script.js (hunk 5 above)
Comment (purple-waves) [reply src/main/webapp/script.js]: Oops, just debugging

Comment (roroross) [reply src/main/webapp/script.js]: we can keep it in for now?  might use it

**Review by roroross — APPROVED** [review]
lgtm

**Review by Techget — APPROVED** [review]
lgtm

**Pull request merged**